introspection-engine/connectors/mongodb-introspection-connector/src/sampler/field_type.rs

```diff
@@ -0,0 +1,107 @@
+use mongodb::bson::Bson;
+use native_types::MongoDbType;
+
+#[derive(Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord)]
+pub(super) enum FieldType {
+    String,
+    Double,
+    BinData,
+    ObjectId,
+    Bool,
+    Date,
+    Int32,
+    Timestamp,
+    Int64,
+    Decimal,
+    Json,
+    Array(Box<FieldType>),
+    Unsupported(&'static str),
+}
+
+impl FieldType {
+    pub(super) fn from_bson(bson: Bson) -> Option<Self> {
+        match bson {
+            Bson::Double(_) => Some(Self::Double),
+            Bson::String(_) => Some(Self::String),
+            Bson::Array(docs) if docs.is_empty() => None,
+            Bson::Array(mut docs) => Some(Self::Array(Box::new(
+                docs.pop()
+                    .and_then(FieldType::from_bson)
+                    .unwrap_or(Self::Unsupported("Unknown")),
+            ))),
+            Bson::Document(_) => Some(Self::Json),
+            Bson::Boolean(_) => Some(Self::Bool),
+            Bson::RegularExpression(_) => Some(Self::Unsupported("RegularExpression")),
+            Bson::JavaScriptCode(_) => Some(Self::Unsupported("JavaScriptCode")),
+            Bson::JavaScriptCodeWithScope(_) => Some(Self::Unsupported("JavaScriptCodeWithScope")),
+            Bson::Int32(_) => Some(Self::Int32),
+            Bson::Int64(_) => Some(Self::Int64),
+            Bson::Timestamp(_) => Some(Self::Timestamp),
+            Bson::Binary(_) => Some(Self::BinData),
+            Bson::ObjectId(_) => Some(Self::ObjectId),
+            Bson::DateTime(_) => Some(Self::Date),
+            Bson::Symbol(_) => Some(Self::Unsupported("Symbol")),
+            Bson::Decimal128(_) => Some(Self::Decimal),
+            Bson::Undefined => Some(Self::Unsupported("Undefined")),
+            Bson::MaxKey => Some(Self::Unsupported("MaxKey")),
+            Bson::MinKey => Some(Self::Unsupported("MinKey")),
+            Bson::DbPointer(_) => Some(Self::Unsupported("DbPointer")),
+            Bson::Null => None,
+        }
+    }
+
+    pub(super) fn is_array(&self) -> bool {
+        matches!(self, Self::Array(_))
+    }
+}
+
+impl fmt::Display for FieldType {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        match self {
+            FieldType::String => write!(f, "String"),
+            FieldType::Double => write!(f, "Double"),
+            FieldType::BinData => write!(f, "BinData"),
+            FieldType::ObjectId => write!(f, "ObjectId"),
+            FieldType::Bool => write!(f, "Boolean"),
+            FieldType::Date => write!(f, "Date"),
+            FieldType::Int32 => write!(f, "Int32"),
+            FieldType::Timestamp => write!(f, "Timestamp"),
+            FieldType::Int64 => write!(f, "Int64"),
+            FieldType::Decimal => write!(f, "Decimal"),
+            FieldType::Json => write!(f, "Document"),
+            FieldType::Array(r#type) => write!(f, "Array({})", r#type),
+            FieldType::Unsupported(r#type) => write!(f, "{}", r#type),
```

Comment:
Please don't change now as it's way too nitpicky, but I try to avoid simple `format!()` and `write!()` calls (look how much code the macro produces, with the tools > expand macros: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=f90f441242167da86e655f6fba8e32e3), here `f.write_str("Document")` or similar would work.

Reply:
I'll change it right away.